quinn/src/incoming_connection.rs

```diff
@@ -0,0 +1,163 @@
+use std::{
+    fmt,
+    future::{Future, IntoFuture},
+    net::{IpAddr, SocketAddr},
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+use bytes::BytesMut;
+use proto::ConnectionError;
+use thiserror::Error;
+
+use crate::{
+    connection::{Connecting, Connection},
+    endpoint::EndpointRef,
+};
+
+/// An incoming connection for which the server has not yet begun its part of the handshake
+#[must_use = "futures/streams/sinks do nothing unless you `.await` or poll them"]
+pub struct IncomingConnection(Option<State>);
+
+struct State {
+    inner: proto::IncomingConnection,
+    endpoint: EndpointRef,
+    response_buffer: BytesMut,
+}
+
+impl IncomingConnection {
+    pub(crate) fn new(
+        inner: proto::IncomingConnection,
+        endpoint: EndpointRef,
+        response_buffer: BytesMut,
+    ) -> Self {
+        Self(Some(State {
+            inner,
+            endpoint,
+            response_buffer,
+        }))
+    }
+
+    /// The local IP address which was used when the peer established
+    /// the connection
+    pub fn local_ip(&self) -> Option<IpAddr> {
+        self.0.as_ref().unwrap().inner.local_ip()
+    }
+
+    /// The peer's UDP address
+    pub fn remote_address(&self) -> SocketAddr {
+        self.0.as_ref().unwrap().inner.remote_address()
+    }
+
+    /// Whether the socket address that is initiating this connection has been validated
+    ///
+    /// This means that the sender of the initial packet has proved that they can receive traffic
+    /// sent to `self.remote_address()`.
+    pub fn remote_address_validated(&self) -> bool {
+        self.0.as_ref().unwrap().inner.remote_address_validated()
+    }
+
+    /// Attempt to accept this incoming connection (an error may still occur)
+    pub fn accept(mut self) -> Result<Connecting, ConnectionError> {
+        let state = self.0.take().unwrap();
+        state
+            .endpoint
+            .accept(state.inner, state.response_buffer)
+            .ok_or_else(|| {
+                ConnectionError::TransportError(proto::TransportError {
+                    code: proto::TransportErrorCode::PROTOCOL_VIOLATION,
+                    frame: None,
+                    reason: "Problem with initial packet".to_owned(),
```

Comment:
This isn't necessarily true. For example, `accept` can fail if the connection limit is exceeded, which isn't a transport error ("The peer violated the QUIC specification") at all. We should propagate an accurate error from the proto layer.

We could overload `LocallyClosed` for the limit-exceeded case, or add a new variant, or introduce an `AcceptError` that's either a `ConnectionError` or a limit-exceeded error. I lean towards a new variant as the clearest and most self-contained approach.

Reply:
New `ConnectionError::ConnectionLimitExceeded` variant added. Error when accepting `IncomingConnection` now originates in quinn-proto, rather than being synthesized in the quinn wrapper.

Return type of `quinn_proto::Endpoint::accept` changed from:

```rs
Result<(ConnectionHandle, Connection), Option<Transmit>>
```

To:

```rs
Result<(ConnectionHandle, Connection), (ConnectionError, Option<Transmit>)>
```